ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/ConnectorUtils.java

```diff
@@ -54,91 +58,112 @@ public static RemoteInferenceInputDataSet processInput(MLInput mlInput, Connecto
-            Map<String, String> processedParameters = new HashMap<>();
-            for (String key : parametersMap.keySet()) {
-                try {
-                    AccessController.doPrivileged((PrivilegedExceptionAction<Void>) () -> {
-                        if (parametersMap.get(key) instanceof String) {
-                            processedParameters.put(key, (String) parametersMap.get(key));
-                        } else {
-                            processedParameters.put(key, gson.toJson(parametersMap.get(key)));
-                        }
-                        return null;
-                    });
-                } catch (PrivilegedActionException e) {
-                    throw new RuntimeException(e);
-                }
-            }
-            inputData = RemoteInferenceInputDataSet.builder().parameters(processedParameters).build();
+            inputData = processTextDocsInput((TextDocsInputDataSet) mlInput.getInputDataset(), connector, parameters, scriptService);
         } else if (mlInput.getInputDataset() instanceof RemoteInferenceInputDataSet) {
             inputData = (RemoteInferenceInputDataSet)mlInput.getInputDataset();
         } else {
             throw new IllegalArgumentException("Wrong input type");
         }
         if (inputData.getParameters() != null) {
             Map<String, String> newParameters = new HashMap<>();
-            inputData.getParameters().entrySet().forEach(entry -> {
-                if (entry.getValue() == null) {
-                    newParameters.put(entry.getKey(), entry.getValue());
-                } else if (StringUtils.isJson(entry.getValue())) {
+            inputData.getParameters().forEach((key, value) -> {
+                if (value == null) {
+                    newParameters.put(key, null);
+                } else if (org.opensearch.ml.common.utils.StringUtils.isJson(value)) {
                     // no need to escape if it's already valid json
-                    newParameters.put(entry.getKey(), entry.getValue());
+                    newParameters.put(key, value);
                 } else {
-                    newParameters.put(entry.getKey(), escapeJson(entry.getValue()));
+                    newParameters.put(key, escapeJson(value));
                 }
             });
             inputData.setParameters(newParameters);
         }
         return inputData;
     }
+    private static RemoteInferenceInputDataSet processTextDocsInput(TextDocsInputDataSet inputDataSet, Connector connector, Map<String, String> parameters, ScriptService scriptService) {
+        List<String> docs = new ArrayList<>(inputDataSet.getDocs());
+        Optional<ConnectorAction> predictAction = connector.findPredictAction();
+        if (predictAction.isEmpty()) {
+            throw new IllegalArgumentException("no predict action found");
+        }
+        String preProcessFunction = predictAction.get().getPreProcessFunction();
+        if (preProcessFunction == null) {
```

Comment:
I assume end user doesn't need to provide this `preProcessFunction`? By default it will be attached to the predictAction?

Reply:
Good finding, this can be attached to the predictAction by default.